ext/pdo_firebird/firebird_driver.c

```diff
@@ -460,16 +460,66 @@ int preprocess(const zend_string* sql, char* sql_out, HashTable* named_params)
 }
 
 /* map driver specific error message to PDO error */
-void _firebird_error(pdo_dbh_t *dbh, pdo_stmt_t *stmt, char const *file, zend_long line) /* {{{ */
+void _firebird_error(pdo_dbh_t *dbh, pdo_stmt_t *stmt, const char *state, const char *msg) /* {{{ */
 {
 	pdo_error_type *const error_code = stmt ? &stmt->error_code : &dbh->error_code;
+	pdo_firebird_db_handle *H = (pdo_firebird_db_handle *)dbh->driver_data;
+	pdo_firebird_error_info *einfo = &H->einfo;
+	int sqlcode = -999;
+
+	if (einfo->errmsg) {
+		pefree(einfo->errmsg, dbh->is_persistent);
+		einfo->errmsg = NULL;
+		einfo->errmsg_length = 0;
+	}
+
+	if (H->isc_status && (H->isc_status[0] == 1 && H->isc_status[1] > 0)) {
+		char buf[512];
+		size_t buf_size = sizeof(buf), read_len = 0;
+		ssize_t tmp_len;
+		const ISC_STATUS *s = H->isc_status;
+		sqlcode = isc_sqlcode(s);
+
+		while ((buf_size > (read_len + 1)) && (tmp_len = fb_interpret(&buf[read_len], (buf_size - read_len - 1), &s)) && tmp_len > 0) {
+			read_len += tmp_len;
+			buf[read_len++] = ' ';
+		}
 
-	strcpy(*error_code, "HY000");
+		/* remove last space */
+		if (read_len) {
+			buf[read_len--] = '\0';
+		}
+
+		einfo->errmsg_length = read_len;
+		einfo->errmsg = pestrndup(buf, read_len, dbh->is_persistent);
+
+#if FB_API_VER >= 25
+		char sqlstate[sizeof(pdo_error_type)];
+		fb_sqlstate(sqlstate, H->isc_status);
+		if (sqlstate != NULL && strlen(sqlstate) < sizeof(pdo_error_type)) {
+			strcpy(*error_code, sqlstate);
+			goto end;
+		}
+#endif
+	} else if (msg) {
+		einfo->errmsg_length = strlen(msg);
```

Comment:
It seems that the msg argument is always constant at the call site. Precomputing the length at call site and passing it as an argument should allow the compiler to precompute it instead of needing a runtime strlen() in the function.

Reply:
Changed to require length in argument

Reply:
Maybe you meant writing the numbers directly in the code without using strlen at all? I may have misunderstood something.